domain/agentbinary/service/store.go

```diff
@@ -84,32 +95,56 @@ func (s *AgentBinaryStore) Add(
 	if err := version.Validate(); err != nil {
 		return errors.Errorf("agent version %q is not valid: %w", version, err)
 	}
+	return s.add(ctx, r, version, size, sha384)
+}
 
+func (s *AgentBinaryStore) add(
+	ctx context.Context, r io.Reader,
+	version coreagentbinary.Version,
+	size int64, sha384 string,
+) (resultErr error) {
 	objectStore, err := s.objectStoreGetter.GetObjectStore(ctx)
 	if err != nil {
 		return errors.Errorf("getting object store: %w", err)
 	}
 
 	path := generatePath(version, sha384)
-	// The object store ignores the already exist error, and always overwrites the object for the given path.
-	// We just need to save the new object store UUID to the database.
 	uuid, err := objectStore.PutAndCheckHash(ctx, path, r, size, sha384)
-	if err != nil {
-		return errors.Errorf("putting agent binary %q: %w", path, err)
+	if err != nil && !errors.Is(err, objectstoreerrors.ErrHashAndSizeAlreadyExists) {
+		return errors.Errorf("putting agent binary of %q with hash %q in the object store: %w", version, sha384, err)
 	}
-	defer func() {
-		if resultErr == nil {
-			return
+	if errors.Is(err, objectstoreerrors.ErrHashAndSizeAlreadyExists) {
+		// This means that the binary already exists in the object store.
+		// Just in case this object is not stored in the agent binary store
+		// table yet because of a previous error, we need to get the UUID of
+		// the existing binary and store it.
+		existingObjectUUID, err := s.st.GetObjectUUID(ctx, path)
+		if err != nil {
+			return errors.Errorf("getting object store UUID for %q: %w", path, err)
 		}
```

Comment:
I think this is fine for now, once we have a better way (e.g. PutAndCheckHash returning the object id with ErrHashAndSizeAlreadyExists) then we can move to that.